lib/iris/experimental/stratify.py

```diff
@@ -0,0 +1,179 @@
+
+from functools import partial
+
+import numpy as np
+import stratify
+
+from iris.cube import Cube
+from iris.coords import Coord, AuxCoord, DimCoord
+
+
+def _copy_coords_without_dim(cube, result, z_dim):
+    """
+    Helper function to copy across non-z coordinates between cubes.
+
+    Parameters
+    ----------
+    cube : iris.cube.Cube instance
+        Incoming cube containing the coordinates to be copied.
+    result : iris.cube.Cube instance
+        Incoming cube for the coordinates to be copied to.
+    z_dim : Dimension within the cube of the z-dimension
+        The dimension will not be copied, as the incoming z-dimension
+        is model level_height, whilst the outgoing z-dimension is pressure.
+
+    Returns
+    -------
+    Cube
+        A single cube with the non-z dimensions added to
+        incoming cube.
+
+    """
+    # Copy across non-Z coordinates.
+    for coord in cube.dim_coords:
+        dim, = cube.coord_dims(coord)
+        if dim != z_dim:
+            result.add_dim_coord(coord.copy(), dim)
+    for coord in cube.aux_coords:
+        dims = cube.coord_dims(coord)
+        if z_dim not in dims:
+            result.add_aux_coord(coord.copy(), dims)
+    for coord in cube.derived_coords:
+        dims = cube.coord_dims(coord)
+        if z_dim not in dims:
+            result.add_aux_coord(coord.copy(), dims)
+
+
+def relevel(phenom, level_data, target_levels, interp_direction=0,
+            interpolator=None):
+    """
+    Interpolate the given phenomenon onto the given levels of level_data.
+
+    For example, suppose we have two datasets `P(i,j,k)` and `H(i,j,k)`
+    and we want ``P(i,j,H)`` we call relevel with
+    ``phenom=P`` and ``level_data=H`` and ``target_levels`` being
+    an array of the values of ``H`` we would like.
+
+    This routine is especially useful for computing isosurfaces of
+    phenomenon that are generally monotonic in the direction of interpolation
+    (such as height/pressure and salinity/depth).
+
+    Parameters
+    ----------
+    phenom : :class:`iris.cube.Cube`
+        The data to be re-levelled.
+
+    level_data : :class:`~iris.cube.Cube` or :class:`~iris.coord.Coord`
+        The coordinate values (cube/coord) in the same system as
+        ``target levels``.
+        All dimensions of ``level_data`` must be broadcastable to ``phenom``
+
+    target_levels : array-like
+        The levels of ``level_data`` to pick from ``phenom``.
+        For the default interpolator, this must be a 1d array.
+        For a custom interpolator this may be multi-dimensional and must
+        broadcast to phenom and level_data except in the phenomenon's
+        ``interp_direction`` dimension.
+
+    interp_direction : int or :class:`~iris.coords.Coord` instance
+        The direction of interpolation. This is necessary because
+        level_data is rarely 1D. This is often the z dimension coordinate
+        of ``phenom``.
```

Comment:
@pelson Needs a re-write if you agree with this [comment](https://github.com/SciTools/iris/pull/2320/files#r103902379).

Also, if a coordinate is provided, it must be 1d or `relevel` will complain unceremoniously with a stack-trace.